### PR TITLE
Engine: version logic on replicas should not be hard coded

### PR DESCRIPTION
**Pull request by bleskes**
The refactoring in #23711 hardcoded version logic for replica to assume monotonic versions. Sadly that's wrong for `FORCE` and `VERSION_GTE`. Instead we should use the methods in VersionType to detect conflicts.

Note - once replicas use sequence numbers for out of order delivery, this logic goes away.

### CODE REVIEW

File: core/src/test/java/org/elasticsearch/index/engine/InternalEngineTests.java [hunk 1]
```diff
@@ -1322,14 +1323,30 @@ public void testVersioningCreateExistsException() throws IOException {
...
+            final long version;
+            switch (versionType) {
+                case INTERNAL:
+                    version = forReplica ? i : Versions.MATCH_ANY;
+                    break;
+                case EXTERNAL:
+                    version = i;
+                    break;
+                case EXTERNAL_GTE:
+                    version = randomBoolean() && i > 0 ? i - 1 : i;
```
Comment (jasontedor) [inline core/src/test/java/org/elasticsearch/index/engine/InternalEngineTests.java]: Is this clearer: `version = randomBoolean() ? Math.max(i - 1, 0) : i;`?

**Review by jasontedor — APPROVED** [review]
LGTM. I left one comment. Is the title of the pull request/commit message correct?

File: core/src/test/java/org/elasticsearch/index/engine/InternalEngineTests.java (hunk 1 above)
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/index/engine/InternalEngineTests.java]: sure/

**Comment by bleskes** [comment]
> Is the title of the pull request/commit message correct?

No... it missed a `not`. Thanks!

**Pull request merged**